### PR TITLE
Use custom transformer when building solution references

### LINKED ISSUE #1025: Custom transformers not working for referenced projects
Hi. I am trying to build a mono-repo project consisting from webpack demo application and separately published library. The library is defined as referenced typescript project (`projectReferences: true`).
In order to build the library correctly I need custom transformer ([@zerollup/ts-transform-paths](https://www.npmjs.com/package/@zerollup/ts-transform-paths) to transform paths). It seems to be ignored when building the library.

### Expected Behaviour

Library code gets emitted while being properly transformed.

### Actual Behaviour

Library code gets emitted but no transformation is applied. It is worth mentioning that `@zerollup/ts-transform-paths` run through `ttypescript` is working fine.

### Steps to Reproduce the Problem

*ts-loader options*

```
{
  projectReferences: true,
  getCustomTransformers: require.resolve('./webpack.tsTransformers'),
}
```

*./webpack.tsTransformers.js*

```
const createTsTransformPath = require('@zerollup/ts-transform-paths')

function getCustomTransformers(program) {
  const tsTransformPath = createTsTransformPath(program)

  return {
    before: [tsTransformPath.before],
    // after: [tsTransformPath.afterDeclarations],
  }
}

module.exports = getCustomTransformers
```

### Location of a Minimal Repository that Demonstrates the Issue.

https://github.com/adam-stanek/chobot/commit/8bf5dd2c69ce0437494d55ac7e9524b0f60c6469

### PR DESCRIPTION
**Pull request by feosuna1**
This PR adds support for custom transformers in project references.

While attempting to use custom transformers, I realized that the transformers were only being applied to the main target and not to any of my referenced projects.  After a bit of spelunking, I found that #1025 was a similar issue to my own.  #1274 attempted to solve the problem, but some changes needed to be made in Typescript to allow for custom transformers in project references. All the necessary changes to support are available in Typescript v4.3+.

I've modeled my changes after @gasnier's original PR (#1025). This change set:
1. Moves the `getCustomTransformers` logic out of `initializeInstance` and into a dedicated function, so that it can be shared.
2. Refactors `makeSolutionBuilderHost` to include the new `getCustomTransformer` field in the `solutionBuilderHost` and have it return the transformers made available in the loader options.

Fixes #1025

### CODE REVIEW

**Comment by johnnyreilly** [comment]
Thanks for this - little snowed right now. Will look when I can

File: src/servicesHost.ts [hunk 1]
```diff
@@ -789,12 +794,28 @@ export function makeSolutionBuilderHost(
       );
       instance.typeReferenceResolutionCache?.update(instance.compilerOptions);
       instance.moduleResolutionCache?.update(instance.compilerOptions);
+
+      if (options) {
+        // The `configFilePath` is the same value that is used as the `project` parameter of
+        // `getCustomtransformers` below.
+        const project = options.configFilePath?.toString();
```
Comment (johnnyreilly) [inline src/servicesHost.ts:801]: is the `toString` necessary?

Comment (johnnyreilly) [reply src/servicesHost.ts:801]: isn't it always a string?

**Comment by johnnyreilly** [comment]
Have updated your branch - it's possible you might need to do some repairs

File: src/servicesHost.ts (hunk 1 above)
Comment (feosuna1) [reply src/servicesHost.ts:801]: You're right.  I don't know what I was thinking.  It's a string.

**Comment by johnnyreilly** [comment]
Failing tests following merge I'm afraid - could you take a look please?

**Comment by feosuna1** [comment]
> Failing tests following merge I'm afraid - could you take a look please?

It started failing because the typescript compiler is complaining that options can be a bunch of different types.

CleanShot 2022-11-26 at 09 24 22@2x

**Comment by johnnyreilly** [comment]
I think we might be nearly there. Could you update the version in the package.json and put an entry in the changeling.md please?

**Review by johnnyreilly — APPROVED** [review]

**Comment by johnnyreilly** [comment]
Thanks!

**Comment by johnnyreilly** [comment]
https://github.com/TypeStrong/ts-loader/releases/tag/v9.4.2

**Pull request merged**